src/main/drivers/at32/dshot_bitbang_stdperiph.c

```diff
@@ -0,0 +1,292 @@
+}
+
+#ifdef USE_DSHOT_TELEMETRY
+void bbSwitchToInput(bbPort_t *bbPort)
+{
+    dbgPinHi(1);
+
+    // Set GPIO to input
+
+    ATOMIC_BLOCK(NVIC_PRIO_TIMER) {
+        MODIFY_REG(bbPort->gpio->cfgr, bbPort->gpioModeMask, bbPort->gpioModeInput);
+    }
+
+    // Reinitialize port group DMA for input
+
+    dmaResource_t *dmaResource = bbPort->dmaResource;
+#ifdef USE_DMA_REGISTER_CACHE
+    bbLoadDMARegs(dmaResource, &bbPort->dmaRegInput);
+#else
+    xDMA_DeInit(dmaResource);
+    xDMA_Init(dmaResource, &bbPort->inputDmaInit);
+    // Needs this, as it is DeInit'ed above...
+    xDMA_ITConfig(dmaResource, DMA_IT_TC, ENABLE);
+#endif
+
+    // Reinitialize pacer timer for input
+
+    bbPort->timhw->tim->cval = 0;
+    bbPort->timhw->tim->pr = bbPort->inputARR;
+
+    bbDMA_Cmd(bbPort, TRUE);
+
+    bbPort->direction = DSHOT_BITBANG_DIRECTION_INPUT;
+
+    dbgPinLo(1);
+}
+#endif
+
+void bbDMAPreconfigure(bbPort_t *bbPort, uint8_t direction)
+{
+    DMA_InitTypeDef *dmainit = (direction == DSHOT_BITBANG_DIRECTION_OUTPUT) ?  &bbPort->outputDmaInit : &bbPort->inputDmaInit;
+
+    dma_default_para_init(dmainit);
+
+    dmainit->loop_mode_enable = FALSE;
+    // dmainit->DMA_Channel = bbPort->dmaChannel;
+    dmainit->peripheral_inc_enable = FALSE;
+    dmainit->memory_inc_enable = TRUE;
+    /* dmainit->DMA_FIFOMode = DMA_FIFOMode_Enable ;
+    dmainit->DMA_FIFOThreshold = DMA_FIFOThreshold_1QuarterFull ;
+    dmainit->DMA_MemoryBurst = DMA_MemoryBurst_Single ;
+    dmainit->DMA_PeripheralBurst = DMA_PeripheralBurst_Single; */
+
+    if (direction == DSHOT_BITBANG_DIRECTION_OUTPUT) {
+        dmainit->priority = DMA_PRIORITY_VERY_HIGH;
+        dmainit->direction = DMA_DIR_MEMORY_TO_PERIPHERAL;
+        dmainit->buffer_size = bbPort->portOutputCount;
+        dmainit->peripheral_base_addr = (uint32_t)&bbPort->gpio->scr;
+        dmainit->peripheral_data_width = DMA_PERIPHERAL_DATA_WIDTH_WORD;
+        dmainit->memory_base_addr = (uint32_t)bbPort->portOutputBuffer;
+        dmainit->memory_data_width = DMA_MEMORY_DATA_WIDTH_WORD;
+
+#ifdef USE_DMA_REGISTER_CACHE
+        xDMA_Init(bbPort->dmaResource, dmainit);
+        bbSaveDMARegs(bbPort->dmaResource, &bbPort->dmaRegOutput);
+#endif
+    } else {
+        dmainit->priority = DMA_PRIORITY_VERY_HIGH;
+        dmainit->direction = DMA_DIR_PERIPHERAL_TO_MEMORY;
+        dmainit->buffer_size = bbPort->portInputCount;
+
+        dmainit->peripheral_base_addr = (uint32_t)&bbPort->gpio->idt;
+
+        dmainit->peripheral_data_width = DMA_PERIPHERAL_DATA_WIDTH_HALFWORD;
+        dmainit->memory_base_addr = (uint32_t)bbPort->portInputBuffer;
+        dmainit->memory_data_width = DMA_MEMORY_DATA_WIDTH_WORD;
+
+#ifdef USE_DMA_REGISTER_CACHE
+        xDMA_Init(bbPort->dmaResource, dmainit);
+        bbSaveDMARegs(bbPort->dmaResource, &bbPort->dmaRegInput);
+#endif
+    }
+}
+
+void bbTIM_TimeBaseInit(bbPort_t *bbPort, uint16_t period)
+{
+    /*TIM_TimeBaseInitTypeDef *init = &bbPort->timeBaseInit;
```

Comment:
Are these comments for testing? or for review?

Reply:
yes, i left them so we can see the stm32 implementation, once this gets a PR, i will remove them